components/application-mgt/org.wso2.carbon.identity.application.mgt/src/test/java/org/wso2/carbon/identity/application/mgt/DefaultRoleManagementListenerTest.java

```diff
@@ -0,0 +1,259 @@
+                    thenReturn(application);
+            when(applicationManagementService.getAllowedAudienceForRoleAssociation(anyString(), anyString())).
+                    thenReturn(APPLICATION_AUD);
+
+            // Calling the preAddRole method.
+            defaultRoleManagementListener.preAddRole(ROLE_NAME, new ArrayList<>(), new ArrayList<>(),
+                    new ArrayList<>(), APPLICATION_AUD, APPLICATION_RES_ID, TENANT_DOMAIN);
+
+            Map<String, Object> threadLocalProps = IdentityUtil.threadLocalProperties.get();
+            if (isFragmentApp) {
+                // If the application is a fragment app, the we add the IS_FRAGMENT_APP property to the thread local.
+                assertTrue(threadLocalProps.containsKey(IS_FRAGMENT_APP));
+            } else {
+                assertFalse(threadLocalProps.containsKey(IS_FRAGMENT_APP));
+            }
+        }
+    }
+
+    @Test(dependsOnMethods = "testPreAddRoleForFragmentApp")
+    public void testPreAddRoleWhenPreviousFragmentPropertyNotCleared() throws Exception {
+
+        try (MockedStatic<ApplicationManagementService> applicationManagementServiceMockedStatic =
+                     mockStatic(ApplicationManagementService.class)) {
+
+            // Creating service provider object.
+            ServiceProvider serviceProvider = createServiceProvider();
+
+            // Creating service provider property object.
+            ServiceProviderProperty isFragmentAppSpProp = buildServiceProviderProperty(IS_FRAGMENT_APP,
+                    Boolean.TRUE.toString());
+            serviceProvider.setSpProperties(new ServiceProviderProperty[]{isFragmentAppSpProp});
+
+            ApplicationManagementService applicationManagementService = mock(ApplicationManagementService.class);
+            applicationManagementServiceMockedStatic.when(ApplicationManagementService::getInstance).
+                    thenReturn(applicationManagementService);
+            when(applicationManagementService.getApplicationByResourceId(anyString(), anyString())).
+                    thenReturn(serviceProvider);
+            when(applicationManagementService.getAllowedAudienceForRoleAssociation(anyString(), anyString())).
+                    thenReturn(APPLICATION_AUD);
+
+            // Validating that the previously added thread local property is still available in the thread. From the
+            // code the existing property will be removed and the new property will be added.
+            Map<String, Object> prevThreadLocalProps = IdentityUtil.threadLocalProperties.get();
+            assertTrue(prevThreadLocalProps.containsKey(IS_FRAGMENT_APP));
+            
+            // Calling the preAddRole method.
+            defaultRoleManagementListener.preAddRole(ROLE_NAME, new ArrayList<>(), new ArrayList<>(),
+                    new ArrayList<>(), APPLICATION_AUD, APPLICATION_RES_ID, TENANT_DOMAIN);
+
```

Comment:
Need to assert that no matter the value is avaialble in the thread local before execute, after the method if the app is a main app, no thread local available / if the app is shared thread local is there.

Reply:
changed